File: src/com/kingmang/lazurite/parser/AST/Statements/DestructuringAssignmentStatement.java
package com.kingmang.lazurite.parser.AST.Statements;

import com.kingmang.lazurite.parser.AST.Expressions.Expression;
import com.kingmang.lazurite.parser.AST.InterruptableNode;
import com.kingmang.lazurite.patterns.visitor.ResultVisitor;
import com.kingmang.lazurite.patterns.visitor.Visitor;
import com.kingmang.lazurite.runtime.Lzr.LzrArray;
import com.kingmang.lazurite.runtime.Lzr.LzrMap;
import com.kingmang.lazurite.core.Types;
import com.kingmang.lazurite.runtime.Value;
import com.kingmang.lazurite.runtime.Variables;
import lombok.AllArgsConstructor;

import java.util.Iterator;
import java.util.List;
import java.util.Map;

@AllArgsConstructor
public final class DestructuringAssignmentStatement extends InterruptableNode implements Statement {
    
    public final List<String> variables;
    public final Expression containerExpression;
    
    @Override
    public void execute() {
        super.interruptionCheck();
        final Value container = containerExpression.eval();
        switch (container.type()) {
            case Types.ARRAY:
                execute((LzrArray) container);
                break;
            case Types.MAP:
                execute((LzrMap) container);
                break;
        }
    }
    
    private void execute(LzrArray array) {
        final int size = variables.size();
        for (int i = 0; i < size; i++) {
            final String variable = variables.get(i);
            if (variable != null) {
                Variables.define(variable, array.get(i));
            }
        }
    }
    private void execute(LzrMap map) {
        int i = 0;
        for (Map.Entry<Value, Value> entry : map) {
            final String variable = variables.get(i);
            if (variable != null) {
                Variables.define(variable, new LzrArray(
                        new Value[] { entry.getKey(), entry.getValue() }
                ));
            }
            i++;
        }
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder();
        sb.append("extract(");
        final Iterator<String> it = variables.iterator();
        if (it.hasNext()) {
            String variable = it.next();
            sb.append(variable == null ? "" : variable);
            while (it.hasNext()) {
                variable = it.next();
                sb.append(", ").append(variable == null ? "" : variable);
            }
        }
        sb.append(") = ").append(containerExpression);
        return sb.toString();
    }
}

File: src/main/java/com/kingmang/lazurite/parser/ast/statements/DestructuringAssignmentStatement.kt
package com.kingmang.lazurite.parser.ast.statements

import com.kingmang.lazurite.core.Types
import com.kingmang.lazurite.parser.ast.InterruptableNode
import com.kingmang.lazurite.parser.ast.expressions.Expression
import com.kingmang.lazurite.patterns.visitor.ResultVisitor
import com.kingmang.lazurite.patterns.visitor.Visitor
import com.kingmang.lazurite.runtime.Variables.define
import com.kingmang.lazurite.runtime.values.LzrArray
import com.kingmang.lazurite.runtime.values.LzrMap

data class DestructuringAssignmentStatement(
    val variables: List<String?>,
    val containerExpression: Expression
) : InterruptableNode(), Statement {
    override fun execute() {
        super.interruptionCheck()
        val container = this.containerExpression.eval()
        when (container.type()) {
            Types.ARRAY -> execute(container as LzrArray)
            Types.MAP -> execute(container as LzrMap)
        }
    }

    private fun execute(array: LzrArray) {
        this.variables.forEachIndexed { i, it -> it?.let { define(it, array[i]) } }
    }

    private fun execute(map: LzrMap) {
        var i = 0
        for ((key, value) in map) {
            val variable = this.variables[i]
            if (variable != null) {
                define(
                    variable, LzrArray(
                        arrayOf(key, value)
                    )
                )
            }
            i++
        }
    }

    override fun accept(visitor: Visitor) =
        visitor.visit(this)

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, input: T): R? =
        visitor.visit(this, input)

    override fun toString(): String {
        val sb = StringBuilder()
        sb.append("extract(")
        val it = this.variables.iterator()
        if (it.hasNext()) {
            var variable = it.next()
            sb.append(variable ?: "")
            while (it.hasNext()) {
                variable = it.next()
                sb.append(", ").append(variable ?: "")
            }
        }
        sb.append(") = ").append(containerExpression)
        return sb.toString()
    }
}
